src/highdicom/seg/sop.py

```diff
@@ -1932,29 +1905,287 @@ def _omit_empty_frames(
+            else:
+                segment_array = pixel_array
+            segment_array = np.around(
+                segment_array * float(max_fractional_value)
+            )
+            segment_array = segment_array.astype(np.uint8)
+        else:
+            if pixel_array.ndim == 3:
+                # "Label maps" that must be converted to binary masks.
+                if number_of_segments == 1:
+                    # We wish to avoid unnecessary comparison or casting
+                    # operations here, for efficiency reasons. If there is only
+                    # a single segment, the label map pixel array is already
+                    # correct
+                    if pixel_array.dtype != np.uint8:
+                        segment_array = pixel_array.astype(np.uint8)
+                    else:
+                        segment_array = pixel_array
+                else:
+                    segment_array = (
+                        pixel_array == segment_number
+                    ).astype(np.uint8)
+            else:
+                segment_array = pixel_array[:, :, :, segment_number - 1]
+                if segment_array.dtype != np.uint8:
+                    segment_array = segment_array.astype(np.uint8)
+
+            # It may happen that a binary valued array is passed that should be
+            # stored as a fractional segmentation. In this case, we also need
+            # to stretch pixel values to 8-bit unsigned integer range by
+            # multiplying with the maximum fractional value.
+            if segmentation_type == SegmentationTypeValues.FRACTIONAL:
+                # Avoid an unnecessary multiplication operation if max
+                # fractional value is 1
+                if int(max_fractional_value) != 1:
+                    segment_array *= int(max_fractional_value)
+
+        return segment_array
+
+    @staticmethod
+    def _get_dimension_index_values(
+        plane_index: int,
+        dimension_position_values: List[np.ndarray],
+        plane_position_values: np.ndarray,
+        has_ref_frame_uid: bool,
+        coordinate_system: Optional[CoordinateSystemNames],
+    ) -> List[int]:
+        """Get dimension index values for a frame.
+
+        Parameters
+        ----------
+        plane_index: int
+            Index of the plane in the sorted planes list.
+        dimension_position_values: List[numpy.ndarray]
+            Relative locations of each plane.
+        plane_position_values: numpy.ndarray
+            Plane positions of each plane.
+        has_ref_frame_uid: bool
+            Whether the source images have a frame of reference.
+        coordinate_system: Optional[highdicom.CoordinateSystemNames]
+            The type of coordinate system used (if any).
+
+        Returns
+        -------
+        index_values: List[int]
+            The dimension index values (except the segment number) for the
+            given plane.
+
```

Comment:
Do we want to reference the `highdicom.seg.DimensionIndexSequence` class? That may help clarify some of the terms used.

Reply:
Addressed in 1774199